### PR TITLE
[DDW-570] Update "View wallet's public key" feature

### PR DESCRIPTION
**Pull request by daniloprates**
This PR updates the "View wallet's public key" feature.

## Todos

- [x] Re-enable functionality in the constant variable
- [x] Implement the password dialog
- [x] Change the field logic to trigger the dialog
- [x] Update the API request
- [x] Enable "view wallet's public key" option for HW 
- [x] Show HW public keys in bech32 format

## Screenshots

Screen Shot 2021-03-03 at 17 49 32
Screen Shot 2021-03-03 at 17 50 08

---

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1614804709195400)
- [ ] Test

## Test Cases

```Gherkin
Scenario 1: Reveal spending password textfield
Given I am on <wallet_type> wallet
When I select reveal button
Then a dialog box appears
And it displays spending password textfield

|wallet_type|
|shelley|


Scenario 2: Reveal public key first time
Given I am on <wallet_type> wallet
And I have correctly entered spending password
When I select reveal wallet public key button
Then wallet public key is displayed in textfield

|wallet_type|
|shelley|


Scenario 3: Reveal public key after successfully displayed the first time
Given I am on <wallet_type> wallet
When I select reveal button
Then the public key is displayed in the textfield

|wallet_type|
|shelley|

Scenario 4: Copy public key
Given public key is displayed in the textfield
When I select copy button
Then public key is copied
And can be pasted correctly

Scenario 5: Show QR code
Given public key is displayed in the textfield
When I select show QR code
Then QR code dialog box is displayed
And wallet public key is displayed

Scenario 6: Hide public key
Given public key is displayed in the textfield
When I select hide button
Then public key is hidden

Scenario 7: Reveal public key after hidden
Given public key is hidden
When I select reveal button
Then public key is displayed in textfield


Scenario 8: Reveal public key first time for hardware wallet
Given I am on hardware wallet
When I select reveal wallet public key button
Then wallet public key is displayed in textfield
```
---

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes
- [ ] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [ ] In English
  - [ ] In Japanese
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

File: source/renderer/app/api/wallets/requests/getWalletPublicKey.js [hunk 1]
```diff
@@ -1,13 +1,18 @@
...
 ): Promise<string> =>
-  request({
-    method: 'GET',
-    path: `/v2/wallets/${walletId}/keys/${role}/${index}`,
-    ...config,
-  });
+  request(
+    {
+      method: 'POST',
+      path: `/v2/wallets/${walletId}/keys/${index}`,
```
Comment (nikolaglumac) [inline source/renderer/app/api/wallets/requests/getWalletPublicKey.js:13]: @daniloprates @aleksandardjordjeviciohk this change has to be reverted. We need both this and the new Api methods as they do different things.

**Review by nikolaglumac — CHANGES_REQUESTED** [review]

File: source/renderer/app/api/wallets/requests/getWalletPublicKey.js (hunk 1 above)
Comment (nikolaglumac) [reply source/renderer/app/api/wallets/requests/getWalletPublicKey.js:13]: We need both the "Create" and "Get" Api methods:
Screenshot 2021-03-04 at 17 59 32

File: source/renderer/app/stores/WalletsStore.js [hunk 2]
```diff
@@ -142,7 +142,9 @@ export default class WalletsStore extends Store {
   /* ----------  Active Wallet  ---------- */
   @observable active: ?Wallet = null;
   @observable activeValue: ?BigNumber = null;
-  @observable activePublicKey: ?string = null;
+  @observable walletsPublicKeys: {
```
Comment (nikolaglumac) [inline source/renderer/app/stores/WalletsStore.js:145]: @aleksandardjordjeviciohk this needs to be reverted. We only keep **one** public key in the store.

File: source/renderer/app/stores/WalletsStore.js [hunk 3]
```diff
@@ -308,24 +311,29 @@ export default class WalletsStore extends Store {
...
     if (!this.active || !IS_WALLET_PUBLIC_KEY_SHARING_ENABLED) {
       return;
     }
 
-    // @TODO Once the api is ready, role and index values should be configured properly
+    // @TODO Once the api is ready, index value should be configured properly
     const walletId = this.active.id;
-    const role = '';
-    const index = '';
+    const index = '1852H';
```
Comment (nikolaglumac) [inline source/renderer/app/stores/WalletsStore.js:325]: @aleksandardjordjeviciohk this needs to be `0H`

File: source/renderer/app/stores/WalletsStore.js [hunk 4]
```diff
@@ -308,24 +311,29 @@ export default class WalletsStore extends Store {
...
+    spendingPassword: passphrase,
+  }: {
+    spendingPassword: string,
+  }) => {
     if (!this.active || !IS_WALLET_PUBLIC_KEY_SHARING_ENABLED) {
       return;
     }
 
-    // @TODO Once the api is ready, role and index values should be configured properly
+    // @TODO Once the api is ready, index value should be configured properly
```
Comment (nikolaglumac) [inline source/renderer/app/stores/WalletsStore.js:323]: @aleksandardjordjeviciohk please remove this comment line

File: source/renderer/app/stores/WalletsStore.js [hunk 5]
```diff
@@ -308,24 +311,29 @@ export default class WalletsStore extends Store {
...
       const walletPublicKey = await this.walletPublicKeyRequest.execute({
         walletId,
-        role,
         index,
+        passphrase,
+        extended,
       }).promise;
       runInAction('update wallet public key', () => {
-        this.activePublicKey = walletPublicKey;
+        this.walletsPublicKeys[walletId] = walletPublicKey;
```
Comment (nikolaglumac) [inline source/renderer/app/stores/WalletsStore.js:336]: @aleksandardjordjeviciohk this needs to be reverted!

File: source/renderer/app/api/wallets/requests/getWalletPublicKey.js (hunk 1 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/api/wallets/requests/getWalletPublicKey.js]: Reverted old one and added new API method.

File: source/renderer/app/stores/WalletsStore.js (hunk 2 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/stores/WalletsStore.js:149]: Reverted

File: source/renderer/app/stores/WalletsStore.js (hunk 3 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/stores/WalletsStore.js]: Updated

File: source/renderer/app/stores/WalletsStore.js (hunk 4 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/stores/WalletsStore.js]: Removed

File: source/renderer/app/stores/WalletsStore.js (hunk 5 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/stores/WalletsStore.js:363]: Reverted

**Comment by nikolaglumac** [comment]
@aleksandardjordjeviciohk please check for which types of wallets do we show this "View wallet's public key" feature.
Is it Shelley only? Or do we show it for Byron wallets and Hardware wallets too?

File: source/renderer/app/stores/WalletsStore.js [hunk 6]
```diff
@@ -266,6 +272,8 @@ export default class WalletsStore extends Store {
     walletsActions.sendMoney.listen(this._sendMoney);
     walletsActions.importWalletFromFile.listen(this._importWalletFromFile);
     walletsActions.chooseWalletExportType.listen(this._chooseWalletExportType);
+    walletsActions.getWalletPublicKey.listen(this._getWalletPublicKey);
```
Comment (nikolaglumac) [inline source/renderer/app/stores/WalletsStore.js:275]: @aleksandardjordjeviciohk please remove this.

File: source/renderer/app/stores/WalletsStore.js [hunk 7]
```diff
@@ -332,6 +339,34 @@ export default class WalletsStore extends Store {
...
+
+    try {
+      const walletPublicKey = await this.getAccountPublicKey.execute({
+        walletId,
+        index,
+        passphrase,
+        extended,
+      }).promise;
+      runInAction('update wallet public key', () => {
+        this.walletsPublicKeys[walletId] = walletPublicKey;
```
Comment (nikolaglumac) [inline source/renderer/app/stores/WalletsStore.js:363]: @aleksandardjordjeviciohk replace with `this.activePublicKey = accountPublicKey;`

File: source/renderer/app/stores/WalletsStore.js [hunk 8]
```diff
@@ -332,6 +339,34 @@ export default class WalletsStore extends Store {
...
+    if (!this.active || !IS_WALLET_PUBLIC_KEY_SHARING_ENABLED) {
+      return;
+    }
+
+    const walletId = this.active.id;
+    const index = '0H';
+    const extended = true;
+
+    try {
+      const walletPublicKey = await this.getAccountPublicKey.execute({
```
Comment (nikolaglumac) [inline source/renderer/app/stores/WalletsStore.js:356]: @aleksandardjordjeviciohk rename to accountPublicKey

Comment (aleksandardjordjeviciohk) [reply source/renderer/app/stores/WalletsStore.js]: Fixed

File: source/renderer/app/stores/WalletsStore.js (hunk 7 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/stores/WalletsStore.js]: Fixed

File: source/renderer/app/stores/WalletsStore.js (hunk 6 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/stores/WalletsStore.js]: Fixed

**Comment by darko-mijic** [comment]
It would make a lot of sense for the dialog to open after we enter a spending password. @daniloprates @nikolaglumac @alexander-rukin?

**Comment by darko-mijic** [comment]
Reveal is not asking for spending password after the key was shown once. Not saying this is wrong, but something we should discuss.

**Comment by nikolaglumac** [comment]
> Reveal is not asking for spending password after the key was shown once. Not saying this is wrong, but something we should discuss.

I believe that is fine. We can easily change it if you want.

**Comment by alexander-rukin** [comment]
> It would make a lot of sense for the dialog to open after we enter a spending password. @daniloprates @nikolaglumac @alexander-rukin?

Agree

**Comment by aleksandardjordjeviciohk** [comment]
> > It would make a lot of sense for the dialog to open after we enter a spending password. @daniloprates @nikolaglumac @alexander-rukin?
> 
> Agree

Added

**Comment by alexander-rukin** [comment]
image
button should be disabled I suppose at this case (while enetered password at least matches our password rules)

**Comment by alexander-rukin** [comment]
image

copy action wrong styles:
now it's always 0.5 opacity for text and 1 for icon 
-> 
it should be 0.7 for both in normal state and 1 for both in hovered

**Comment by alexander-rukin** [comment]
> Reveal is not asking for spending password after the key was shown once. Not saying this is wrong, but something we should discuss.

It's probably OK for one wallet, but now it doesn't ask for password in other wallets too!!!
MOREOVER
Is this OK that all my wallets have the same public key? 

image
image
image

**Comment by miorsufianiohk** [comment]
I am not sure why the buttons 'REVEAL' and 'HIDE' are both in capital letters which seem to be inconsistent with the other buttons.
Screenshot 2021-03-08 at 11 24 20

**Comment by alexander-rukin** [comment]
> I am not sure why the buttons 'REVEAL' and 'HIDE' are both in capital letters which seem to be inconsistent with the other buttons.
> Screenshot 2021-03-08 at 11 24 20

this is by design for special "input buttons"

**Comment by miorsufianiohk** [comment]
> > I am not sure why the buttons 'REVEAL' and 'HIDE' are both in capital letters which seem to be inconsistent with the other buttons.
> 
> this is by design for special "input buttons"

@nikolaglumac @darko-mijic  is this really what we want, a mixture of buttons in entirely uppercase and some entirely lowercase but the first letter? From user point of view, I don't see anything special with 'REVEAL' or 'HIDE' buttons.

**Comment by miorsufianiohk** [comment]
There's a pipe symbol, on the very left hand side of the buttons when 'Reward view' is chosen which is inconsistent with other views. Tested on 17020
List view
List view
Reward view
Reward view
Tile view
Tile view

**Comment by nikolaglumac** [comment]
@alexander-rukin @miorsufianiohk can we perhaps use the 👁️   icon here instead of "REVEAL"?

**Comment by nikolaglumac** [comment]
@miorsufianiohk @alexander-rukin @darko-mijic all issues have been resolved. Please re-test 🙏 

@miorsufianiohk we won't be changing the "REVEAL/HIDE" designs for now so let's take that discussion out of this PR. Thanks!

**Comment by nikolaglumac** [comment]
@darko-mijic one thing to be aware of - the public key returned from the API is in the bech32 format so it looks something like this:

> acct_xvk1cm454leclzk352njpt77qevyuw0ac5k90gzqrztlce8m0a5spg3sq7rd7fsynpe3lkn74a6l5dnmjaf9le5wam88lxcmakctag5hs8gq78hah

...while the one we have for hardware wallets is in hex and looks like:

> 45cc0e2b463831023afcd441b212cdca539ef202b612b59eb1944fbfc8e75b3g

@tomislavhoracek @KtorZ  can we convert those hex values to bech32 format?

Screenshot 2021-03-08 at 18 46 59

**Comment by tomislavhoracek** [comment]
@daniloprates @nikolaglumac Support for Hardware Wallets improved. Now we have `bech32` format for HW xpub.

**Review by nikolaglumac — APPROVED** [review]

**Comment by darko-mijic** [comment]
Reviewing this now...

**Comment by darko-mijic** [comment]
Reveal public key should open the popup for hardware wallets as for software wallets.

**Review by miorsufianiohk — APPROVED** [review]
LGTM. Great job @nikolaglumac 👍

**Comment by gabriela-ponce** [comment]
@daniloprates I found these issues on 17048:

- When syncing hardware wallets a password is requested to reveal the public key. Check [this](https://drive.google.com/file/d/1FK8jvudPO9Fsd7JhdhxKkabJJwVZN9hU/view?usp=sharing) video.
- If you insert a short password the message reads "An error occurred" instead of "Incorrect wallet password". Check [this](https://drive.google.com/file/d/169pBTRavYFKJXVwV_jScjdL0Bl2SFM-m/view?usp=sharing) video.

**Review by miorsufianiohk — CHANGES_REQUESTED** [review]
Hi @daniloprates . On 17052 I have narrowed down the issue that we are facing with spending password dialog box for hw wallet. cc: @nikolaglumac 

I found this:
1. For HW wallet, when I pressed reveal button (Result: Spending password dialog box displayed)
2. I dismissed the spending password dialog box
3. Then, I went to SW wallet, pressed reveal button (Result: Spending password dialog box displayed)
4. I dismissed the spending password dialog box
5. I went back to HW wallet and pressed reveal button (Result: no spending dialog box displayed anymore and public key is populated in textfield)
6. After this going back and forth SW and HW wallets, don't make HW wallets spending password dialog to come up again.

For further info please [watch video](https://drive.google.com/file/d/1dCTueI_GRMlcS1NCbfVLg9J91eVGO1dm/view?usp=sharing)

**Comment by nikolaglumac** [comment]
@miorsufianiohk all these issues have just been fixed.

**Review by miorsufianiohk — APPROVED** [review]
LGTM. Great job @nikolaglumac 👍 . Tested on 17055

**Pull request merged**